File: 0046-permutations/0046-permutations.go
func swap(a *int, b *int) {
    temp := *a
    *a = *b
    *b = temp
}

func solve(idx int, nums []int, res *[][]int) {
    if idx == len(nums)-1 {
        // Create a new slice and copy elements from nums
        permutation := make([]int, len(nums))
        copy(permutation, nums)
        *res = append(*res, permutation)
        return
    }

    for i := idx; i < len(nums); i++ {
        swap(&nums[idx], &nums[i])
        solve(idx+1, nums, res)
        swap(&nums[i], &nums[idx]) // Backtrack by swapping again
    }
}

func permute(nums []int) [][]int {
    var res [][]int
    solve(0, nums, &res)
    return res
}

File: 0046-permutations/0046-permutations.java
class Solution {
    private void backtrack(List<Integer> curr, boolean[] used, List<List<Integer>> res, int[] nums) {
        if(curr.size() == nums.length) {
            res.add(new ArrayList<>(curr));
        }
        for(int i = 0; i < nums.length; i++) {
            if(used[i]) continue;
            used[i] = true;
            curr.add(nums[i]);
            backtrack(curr, used, res, nums);
            curr.remove(curr.size() - 1);
            used[i] = false;
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        backtrack(curr, used, res, nums);
        return res;
    }
}